brave-cxf3/src/main/java/com/github/kristofa/brave/cxf3/TracerContext.java

```diff
@@ -0,0 +1,100 @@
+import com.github.kristofa.brave.ServerSpanThreadBinder;
+import org.apache.cxf.message.Message;
+
+import java.util.concurrent.Callable;
+
+/**
+ * Used to handle {@link ServerSpan} retrieved from cxf message exchange.
+ */
+public class TracerContext {
+
+  private final ServerSpanThreadBinder serverSpanThreadBinder;
+  private final Message message;
+
+  TracerContext(Brave brave, Message message) {
+    this.serverSpanThreadBinder = brave.serverSpanThreadBinder();
+    this.message = message;
+  }
+
+  /**
+   * Wraps provided {@link Runnable and ensures same {@link ServerSpan} context as the thread from which the Runnable was executed.
+   * {@link ServerSpan} is taken from cxf {@link org.apache.cxf.message.Message} {@link org.apache.cxf.message.Exchange}
+   *
+   * @param command {@link Runnable}
+   * @return wrapped {@link Runnable}
+   */
+  public Runnable wrap(Runnable command) {
+    if (command == null) {
+      return null;
+    }
+
+    // get server span bind to current thread
+    final ServerSpan current = getCurrentServerSpan();
+    final ServerSpan serverSpan = getFromExchange();
+
+    return () -> {
+      try {
+        // When current span is not null, it means it was bind manually by developer
+        // or this wrapper is used in sync scenario.
+        // Bind server span from Exchange only when current span is null.
+        if (current == null) {
+          serverSpanThreadBinder.setCurrentSpan(serverSpan);
+        }
+        command.run();
+      } finally {
+        serverSpanThreadBinder.setCurrentSpan(current);
+      }
+    };
+  }
+
+  /**
+   * Wraps provided {@link Callable} and ensures same {@link ServerSpan} context as the thread from which the Callable was executed.
+   * {@link ServerSpan} is taken from cxf {@link org.apache.cxf.message.Message} {@link org.apache.cxf.message.Exchange}
+   *
+   * @param command {@link Callable}
+   * @return wrapped {@link Callable}
+   */
+  public <V> Callable<V> wrap(Callable<V> command) {
+    if (command == null) {
+      return null;
+    }
+
+    // get server span bind to current thread
+    final ServerSpan current = getCurrentServerSpan();
+    final ServerSpan serverSpan = getFromExchange();
+
+    return () -> {
+      try {
+        // When current span is not null, it means it was bind manually by developer
+        // or this wrapper is used in sync scenario.
+        // Bind server span from Exchange only when current span is null.
+        if (current == null) {
+          serverSpanThreadBinder.setCurrentSpan(serverSpan);
+        }
+        return command.call();
+      } finally {
+        serverSpanThreadBinder.setCurrentSpan(current);
+      }
+    };
+  }
+
+  private ServerSpan getCurrentServerSpan() {
+    // gets existing server span or reinitializes to ServerSpan#EMPTY using ThreadLocal#initialValue
+    final ServerSpan current = serverSpanThreadBinder.getCurrentServerSpan();
+
+    // when value was just initialized
```

Comment:
ps I looked into the code, and actually we can't depend on any semantics around null. For example, ServerSpan.EMPTY is indeed the initial value, but both of our implementations revert to this on `setServerSpan(null)` (via threadlocal.remove which reverts to initial value which is EMPTY!)

in #321 I clarified this by explicitly setting to ServerSpan.EMPTY (since that's what ends up happening anyway).

Long story short, we have no means to tell whether something was explicitly set to EMPTY or by accident as eventhough javadoc say don't accept null, we did, and also we don't control all implementations of span state :(